cell_classification/predict.py

```diff
@@ -0,0 +1,236 @@
+import toml
+import argparse
+import pickle
+import os
+from sklearn.metrics import roc_curve, auc, confusion_matrix
+from model_builder import ModelBuilder
+import numpy as np
+from tqdm import tqdm
+from plot_utils import plot_average_roc, plot_metrics_against_threshold
+import toml
+from copy import deepcopy
+from joblib import Parallel, delayed
+import pandas as pd
+
+
+def load_model_and_val_data(params):
+    """Load model and validation data from params dict
+    Args:
+        params (dict):
+            dictionary containing model and validation data paths
+    Returns:
+        model (ModelBuilder):
+            trained model
+        val_data (tf.data.Dataset):
+            validation dataset
+    """
+    params["eval"] = True
+    model = ModelBuilder(params)
+    model.prep_data()
+    model.load_model(params["model_path"])
+    val_dset = iter(model.validation_dataset)
+    return model, val_dset
+
+
+def predict(model, val_dset, params):
+    """Predict labels for validation data
+    Args:
+        model (ModelBuilder):
+            trained model
+        val_dset (tf.data.Dataset):
+            validation data
+        params (dict):
+            dictionary containing model and validation data paths
+    Returns:
+        predictions (list):
+            list of dictionaries containing the full example + prediction
+        params (dict):
+            dictionary containing model hyperparams and validation data paths
+    """
+    pred_list = []
+    for sample in tqdm(val_dset):
+        sample["prediction"] = model.predict(model.prep_batches(sample)[0])
+        pred_list.append(sample)
+
+    # split batches to single samples
+    single_example_list = []
```

Comment:
Is this because the tf_record is constructed with a specific batch_size already?

Reply:
No, the dataloader has a batchsize, the tf_record just stores single examples. I did this because we want to do inference on larger batches. Atm with batchsize=4 my laptop GPU is far from fully utilized and it runs wit 5-7 batches per second.

Reply:
so when you run `model.predict`, that is predicting 4 images at a time. Then the logic from line 56 onwards is splitting that batch back into individual examples?

Reply:
Yes exactly. My rationale was that we'll apply this model on larger images anyways and need to do tile & stitch for prediction, so it'll almost always run on a batch of tiles and not on single tiles.